### ISSUE TITLE
LDAP Group sync issue with AEM 6.2

### ISSUE THREAD

**Issue by waringme** [Netcentric/accesscontroltool#140]
1. clean instance on aem 6.2 SP 1 
2. created groups with netcentric tool and deployed. for example there is a group -AEM_SYSYTEM_ADMIN
3. same groups setup in ldap AEM_SYSYTEM_ADMIN.
4. user1 is in ldap and is a member of AEM_SYSYTEM_ADMIN.
5. user1 get authenticated but fails to get authorisation.  AEM sees the AEM_SYSYTEM_ADMIN as a local group.
6. Deleted the local AEM_SYSYTEM_ADMIN.
7. user1 logs in again and the group from ldap is synced acrossed. because the groups has already had already been run in step 2. false positive.
8. delete the synced groups from crx.
9. installed the netcentric group package again so the AEM_SYSYTEM_ADMIN is created.
10. logged in again with user1.
11. netcentric group AEM_SYSYTEM_ADMIN is now showing as synced with ldap and displaying the members (user 1 etc)

Is there a way to controlling permissions for LDAP groups? 

Thanks

Tim

**Comment by ghenzler** [comment]
@waringme Can you check in more detail what the group AEM_SYSTEM_ADMIN looks like in JCR when created by AC Tool vs. created by LDAP sync? I could imagine that a flag was introduced with SP1 (SAML has such a flag for external groups). Also, can you please provide your oak version? (current oak hotfixes have changed LDAP behaviour, SP1 has oak 1.4.6 [2])

[1] https://helpx.adobe.com/experience-manager/kb/aem62-available-hotfixes/OakCumulativeFixPack.html#Oak147
[2] https://docs.adobe.com/content/docs/en/aem/6-2/release-notes/sp1.html#contentbody_text_2013062266

**Comment by waringme** [comment]
I've not got access to the system at the moment.  from memory the external groups nodes looked very similar to 6.1. I can check next Monday. 
Version:  oak 1.4.9.
Server setup was done by
1. AEM 6.2 GA
2. Hotfix 12785
3. SP 1
4 .Hot fixes 12444, 13062,13547

**Comment by ghenzler** [comment]
The root cause for it is really the change in OAK-4397 (change active since oak 1.4.7), possible workarounds are described in [1]. 

For AC Tool users this means the easiest workaround is to always run ldap sync via JMX *before* running the AC Tool (hence new groups would always be created by the LDAP module and therefore the problem should not show up). 

NOTE: The ldap module creates groups where rep:authorizableId and rep:principalName differ  (e.g. `group-name` for `rep:authorizableId` and `cn=group-name,ou=mydepart,ou=Groups,dc=comp,dc=com;IDPNAME` for `rep:principalName`). This means that when the AC Tool uses existing LDAP groups it will create ACEs with with a long name like `cn=group-name,ou=mydepart,ou=Groups,dc=comp,dc=com;IDPNAME` which looks wrong but is correct and works (as the principal name is the one to be referenced from ACEs). 

To improve the ease of use for the AC Tool, the following options exist:

**1. Provide a config to run the JMX LDAP sync from AC Tool**

This way it could be ensured that groups are created by LDAP module, however the sync potentially could run a long time and introduce unnecessary overhead. 

**2. Extend existing `allowExternalGroupNamesRegEx` to allow to explicitly mention LDAP groups**

This property can currently be used to not delete members from AC Tool config groups. However currently it is not possible yet to specify the following:

```
- global_config:
      allowExternalGroupNamesRegEx: LDAPPREFIX.*

- group_config:

    - aem-only-editor-group:
        - name: "AEM Role Editor"
          members: LDAPPREFIX-AEM-EDITORS
```

As mentioned in [1]:

> Also, in the general principal and access control design, the external group might be used solely for the representation of the memberships. A separate local "role" group would be used for defining the access control, and this one would include the external group as member, so the permissions get inherited to the external users that are member of that group.

Even though this option introduces some overhead regarding the number of groups (all groups have to exist twice, once in the "LDAP version" and once in the "AEM role version", for large customers this can be a large magnitude), this avoids have DN entries in the content ACEs (see note from above) and provides a more loosely-coupled way to integrate with LDAP.

**3. Introduce attribute `externalId` to group configurations **

This could be used specify the external Id in the AC configuration. Since the LDAP module is based on oak-auth-external, there is an API to create groups in the exact same way as the LDAP module would [3]. However this introduces strong coupling to the ldap module.

[1]
https://issues.apache.org/jira/browse/OAK-4397?focusedCommentId=15556027&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15556027

[2]
https://github.com/Netcentric/accesscontroltool/blob/develop/docs/AdvancedFeatures.md#configure-memberships-to-dynamic-groups

[3]
https://github.com/apache/jackrabbit-oak/blob/jackrabbit-oak-1.4.7/oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/basic/DefaultSyncContext.java#L396

**Comment by kwin** [comment]
Here is my feedback regarding the different options:

1. Not really feasible with larger LDAP as syncing all users in advance takes way too much time. Just syncing the groups would be an option, but the according JIRA ticket ([OAK-1823](https://issues.apache.org/jira/browse/OAK-1823)) was closed as "Won't Fix". For a more comprehensive reasoning look at https://issues.apache.org/jira/browse/OAK-1823?focusedCommentId=15262596&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15262596.
2. Might be feasible but the overhead is huge. We already had quite some problems in the past due to too many groups in some systems.
3. A bit nasty, but looks like the best option in this case. In case the group should be created by the ACTool (reasonable in some cases) and still supposed to be used by LDAP users this looks like the most straightforward fix to this issue.

**Comment by gruberroland-netcentric** [comment]
+1 for option 3)
This way only the needed groups are created and not a full LDAP sync is required. There would be no much change required in AC Tool as only the creation of the group would differ.

**Comment by ghenzler** [comment]
@gruberroland-netcentric @kwin The main problem with option three is the correct LDAP dn + IDP Config name would need to be mentioned in the yaml config, otherwise the LDAP module will not pick it up correctly:

```
- group_config:

    - aem-only-editor-group:
        - name: "AEM Role Editor"
          externalId: "cn=group-name,ou=mydepart,ou=Groups,dc=comp,dc=com;IDPNAME"
```
The first part of the external id is the full LDAP DN, the second part ("IDPNAME" in the example) is name of the IDP as configured via PID `org.apache.jackrabbit.oak.security.authentication.ldap.impl.LdapIdentityProvider` - creating configurations like this highly couples the yaml config to the AEM LDAP module config (at least on AEM side, not too bad) but also the LDAP structure on LDAP server side. This means if groups are moved on LDAP side, the AEM setup would not work anymore. Using option 2 would make a lot less assumptions about the LDAP side (only the group name, not the LDAP location). I also see the overhead problem, but for all cases where the overhead of using two groups is bearable, I think I would prefer overhead & loose coupling. 

Also, if going down the path of option three, we probably should introduce variables so at least the string `,ou=mydepart,ou=Groups,dc=comp,dc=com;IDPNAME` could be set at one position and then referenced in the group like `externalId: "cn=group-name,${LDAPSUFFIX}"`, otherwise we would have too much duplication the config file IMHO.

**Comment by gruberroland-netcentric** [comment]
It is not that common to move groups in LDAP because this
* often requires to update ACLs on server side
* applications often depend on a search suffix to find the groups (e.g. ou=Groups,dc=comp,dc=com)
* other LDAP objects might reference the group via DN (e.g. nested groups)

So this would probably happen more likely in case of a merger or fundamental restructuring.

Also the LDAP sync needs to handle this some way.
Either it updates the group when it finds it in new location (then we are fine creating it with old DN).
Or it cannot handle moves in the directory structure. Then there would be an issue if we create the group before the sync. But I would say the risk for this is low.

**Comment by ghenzler** [comment]
Ok, let's implement option 3 for this ticket then.

**Comment by waringme** [comment]
Many thanks for the detailed responses.  
Option 1 =  currently around 100 groups but would also sync the the users, the customer has 100k users in the ldap
Between option 2 and option 3 - whilst I would prefer to decouple from ldap, the groups in ldap are unlikey to move. We do have two ldaps (ldap-au & ldap-nz), would the solution support multiple ${LDAPSUFFIX} ? If it can than option 3 seems the best option for now.

**Comment by waringme** [comment]
Thanks, do you have a ETA when you will be able to update the AC Tool?

**Comment by ghenzler** [comment]
PR #154 introduces an attribute `externalId`. It can be used directly as in https://github.com/Netcentric/accesscontroltool/issues/140#issuecomment-282662165 or using with variables as described in #153 to avoid duplication.

 The fact that for the externalId case `rep:authorizableId` is not equal to `rep:principalName`, revealed a couple of bugs. For instance the ACL part of the dumps did not look up correctly for principal names (but for authorizableId which is wrong). Also principalName was misused as "profile name" which was highly confusing. The AuthorizableConfigBean has now a new property `name` that contains the profile name (and the principalName usuasally is now equal to the authorizableId in this bean as well). Also the round trip (dump and import) was loosing the profile name information, this is fixed in #154 as well.

**Comment by mittalkunal** [comment]
I know the issue is closed but thought of posting this as this is related.

We did run into the same issue and we could not use the above approach as there were a lot of inconsistencies between the values of CN and OU. So below is what we did.

We created AEM groups as well as LDAP groups via Netcentric tool. Added LDAP group as a member to AEM group. Building via Maven, once the module for Netcentric is done, another module takes care of deleting the LDAP groups. 

Whenever LDAP group gets synced within AEM, the weak reference stored within AEM groups remains intact thus the mapping between AEM and LDAP groups work.
